core/src/cgroup.rs

```diff
@@ -1,238 +1,392 @@
+        if !is_cgroup(&self.path)? {
+            bail!("{} is not a valid cgroup", self.path.to_str().unwrap());
+        }
+
+        fs::write(self.path.join("cgroup.procs"), pid.to_string())?;
+        Ok(())
+    }
+
+    /// Returns all PIDs associated with this cgroup
+    pub fn get_pids(&self) -> anyhow::Result<Vec<Pid>> {
+        if !is_cgroup(&self.path)? {
+            bail!("{} is not a valid cgroup", self.path.to_str().unwrap());
+        }
 
+        // TODO: "Rustify" this
+        let data = fs::read(self.path.join("cgroup.procs"))?;
+        let mut pids: Vec<Pid> = Vec::new();
+        for line in data.lines() {
+            let line = line?;
+            pids.push(Pid::from_raw(line.parse()?));
+        }
+
+        Ok(pids)
+    }
+
+    /// Checks whether this cgroup is populated
+    pub fn populated(&self) -> anyhow::Result<bool> {
+        if !is_cgroup(&self.path)? {
+            bail!("{} is not a valid cgroup", self.path.to_str().unwrap());
+        }
+
+        Ok(!fs::read(self.path.join("cgroup.procs"))?.is_empty())
+    }
+
+    /// Checks whether this cgroup is frozen
+    pub fn frozen(&self) -> anyhow::Result<bool> {
+        if !is_cgroup(&self.path)? {
+            bail!("{} is not a valid cgroup", self.path.to_str().unwrap());
+        }
+
+        // We need to check for the existance of cgroup.freeze, because
+        // this file does not exist on the root cgroup.
+        let path = self.path.join("cgroup.freeze");
         if !path.exists() {
-            trace!("Creating Cgroup, {path:?}");
-            std::fs::create_dir(&path).typ(SystemError::CGroup)?;
+            return Ok(false);
+        }
+
+        Ok(fs::read(&path)? == "1\n".as_bytes())
+    }
+
+    /// Freezes this cgroup (does nothing if already frozen)
+    pub fn freeze(&self) -> anyhow::Result<()> {
+        if !is_cgroup(&self.path)? {
+            bail!("{} is not a valid cgroup", self.path.to_str().unwrap());
         }
 
-        // TODO use cpuset with feature opt in
-        //let cont = path.join("cgroup.subtree_control");
-        //println!("{cont:?}");
-        //unsafe {exit(1)};
-        //std::fs::write(path.join("cgroup.subtree_control"), b"+pids")?;
-        //sleep(Duration::from_secs(1));
-        //sleep(Duration::from_secs(1));
+        // We need to check for the existance of cgroup.freeze, because
+        // this file does not exist on the root cgroup.
+        let path = self.path.join("cgroup.freeze");
+        if !path.exists() {
+            bail!("cannot freeze the root cgroup");
+        }
 
-        Ok(CGroup { path })
+        Ok(fs::write(path, "1")?)
     }
 
-    /// Returns a file descriptor to the list of the processes that belong to this cgroup
-    /// TODO: Does this need to be public?
-    pub fn get_procs_fd(&self) -> TypedResult<OwnedFd> {
-        File::open(self.path.join(Self::MEMBER_FILE))
-            .typ(SystemError::CGroup)
-            .map(OwnedFd::from)
```

Comment:
Because I try to keep the API minimal and to my knowledge, FDs are a one-way, meaning that it is possible to create an FD from a path but not obtain a path from an FD due to the fact, that not all FDs are files on FS.